File: email-archive/email_archive/attachment_extractor.py
"""Extract text from email attachments."""

from __future__ import annotations

import io
import os

SUPPORTED_EXTENSIONS = frozenset([".pdf", ".docx", ".xlsx", ".txt", ".text", ".csv"])

EXTENSION_TO_MIME = {
    ".pdf": "application/pdf",
    ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ".doc": "application/msword",
    ".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    ".xls": "application/vnd.ms-excel",
    ".txt": "text/plain",
    ".text": "text/plain",
    ".csv": "text/csv",
}

SUPPORTED_MIMES = frozenset([
    "application/pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "text/plain",
    "text/csv",
])


def get_mime_type(filename: str) -> str:
    """Get MIME type from filename extension."""
    ext = os.path.splitext(filename)[1].lower()
    return EXTENSION_TO_MIME.get(ext, "application/octet-stream")


def can_extract(filename: str, mime_type: str | None = None) -> bool:
    """Check if text can be extracted from this attachment type."""
    ext = os.path.splitext(filename)[1].lower()
    if ext in SUPPORTED_EXTENSIONS:
        return True
    if mime_type and mime_type in SUPPORTED_MIMES:
        return True
    return False


def extract_text(filename: str, content: bytes, mime_type: str | None = None) -> str | None:
    """Extract text from attachment bytes.

    Returns None if extraction fails or yields no text.
    """
    if not content:
        return None
    if not can_extract(filename, mime_type):
        return None

    ext = os.path.splitext(filename)[1].lower()
    effective_mime = mime_type or get_mime_type(filename)

    try:
        if ext == ".pdf" or effective_mime == "application/pdf":
            return _extract_pdf(content)
        elif ext == ".docx" or "wordprocessingml" in effective_mime:
            return _extract_docx(content)
        elif ext == ".xlsx" or "spreadsheetml" in effective_mime:
            return _extract_xlsx(content)
        elif ext in (".txt", ".text", ".csv") or effective_mime.startswith("text/"):
            return _extract_text(content)
        else:
            return None
    except Exception:
        return None


def _extract_pdf(content: bytes) -> str | None:
    import pdfplumber

    with pdfplumber.open(io.BytesIO(content)) as pdf:
        pages = []
        for page in pdf.pages:
            text = page.extract_text()
            if text:
                pages.append(text)
    text = "\n".join(pages).strip()
    return text if text else None


def _extract_docx(content: bytes) -> str | None:
    from docx import Document

    doc = Document(io.BytesIO(content))
    paragraphs = [p.text for p in doc.paragraphs]
    text = "\n".join(paragraphs).strip()
    return text if text else None


def _extract_xlsx(content: bytes) -> str | None:
    from openpyxl import load_workbook

    wb = load_workbook(io.BytesIO(content), read_only=True, data_only=True)
    parts = []
    for ws in wb.worksheets:
        parts.append(f"--- {ws.title} ---")
        for row in ws.iter_rows(values_only=True):
            cells = [str(c) if c is not None else "" for c in row]
            parts.append("\t".join(cells))
    wb.close()
    text = "\n".join(parts).strip()
    return text if text else None


def _extract_text(content: bytes) -> str | None:
    text = content.decode("utf-8", errors="replace").strip()
    return text if text else None

File: email-archive/src/EmailArchive/Ingest/AttachmentExtractor.cs
using System.Text;
using ClosedXML.Excel;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using UglyToad.PdfPig;

namespace EmailArchive.Ingest;

public class AttachmentExtractor
{
    private static readonly HashSet<string> SupportedExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".pdf", ".docx", ".xlsx", ".txt", ".text", ".csv"
    };

    private static readonly Dictionary<string, string> ExtensionToMimeType = new(StringComparer.OrdinalIgnoreCase)
    {
        [".pdf"] = "application/pdf",
        [".docx"] = "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        [".doc"] = "application/msword",
        [".xlsx"] = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        [".xls"] = "application/vnd.ms-excel",
        [".txt"] = "text/plain",
        [".text"] = "text/plain",
        [".csv"] = "text/csv"
    };

    private static readonly HashSet<string> SupportedMimeTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "application/pdf",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "text/plain",
        "text/csv"
    };

    /// <summary>
    /// Check if text can be extracted from this attachment type.
    /// </summary>
    public bool CanExtract(string filename, string? mimeType)
    {
        var extension = Path.GetExtension(filename);

        if (SupportedExtensions.Contains(extension))
            return true;

        if (!string.IsNullOrEmpty(mimeType) && SupportedMimeTypes.Contains(mimeType))
            return true;

        return false;
    }

    /// <summary>
    /// Get MIME type from filename extension.
    /// </summary>
    public string GetMimeType(string filename)
    {
        var extension = Path.GetExtension(filename);
        return ExtensionToMimeType.GetValueOrDefault(extension, "application/octet-stream");
    }

    /// <summary>
    /// Extract text from attachment bytes.
    /// </summary>
    public string? ExtractText(string filename, string? mimeType, byte[] content)
    {
        if (content.Length == 0)
            return null;

        if (!CanExtract(filename, mimeType))
            return null;

        var extension = Path.GetExtension(filename).ToLowerInvariant();
        var effectiveMimeType = mimeType ?? GetMimeType(filename);

        try
        {
            return extension switch
            {
                ".pdf" => ExtractFromPdf(content),
                ".docx" => ExtractFromDocx(content),
                ".xlsx" => ExtractFromXlsx(content),
                ".txt" or ".text" or ".csv" => ExtractFromText(content),
                _ when effectiveMimeType == "application/pdf" => ExtractFromPdf(content),
                _ when effectiveMimeType.Contains("wordprocessingml") => ExtractFromDocx(content),
                _ when effectiveMimeType.Contains("spreadsheetml") => ExtractFromXlsx(content),
                _ when effectiveMimeType.StartsWith("text/") => ExtractFromText(content),
                _ => null
            };
        }
        catch
        {
            // If extraction fails, return null rather than crashing
            return null;
        }
    }

    private string? ExtractFromPdf(byte[] content)
    {
        using var stream = new MemoryStream(content);
        using var document = PdfDocument.Open(stream);

        var sb = new StringBuilder();
        foreach (var page in document.GetPages())
        {
            sb.AppendLine(page.Text);
        }

        var text = sb.ToString().Trim();
        return string.IsNullOrWhiteSpace(text) ? null : text;
    }

    private string? ExtractFromDocx(byte[] content)
    {
        using var stream = new MemoryStream(content);
        using var doc = WordprocessingDocument.Open(stream, false);

        var body = doc.MainDocumentPart?.Document?.Body;
        if (body is null)
            return null;

        var sb = new StringBuilder();
        foreach (var para in body.Elements<Paragraph>())
        {
            sb.AppendLine(para.InnerText);
        }

        var text = sb.ToString().Trim();
        return string.IsNullOrWhiteSpace(text) ? null : text;
    }

    private string? ExtractFromXlsx(byte[] content)
    {
        using var stream = new MemoryStream(content);
        using var workbook = new XLWorkbook(stream);

        var sb = new StringBuilder();
        foreach (var worksheet in workbook.Worksheets)
        {
            sb.AppendLine($"--- {worksheet.Name} ---");

            var usedRange = worksheet.RangeUsed();
            if (usedRange is null)
                continue;

            foreach (var row in usedRange.Rows())
            {
                var cells = row.Cells().Select(c => c.GetString());
                sb.AppendLine(string.Join("\t", cells));
            }
        }

        var text = sb.ToString().Trim();
        return string.IsNullOrWhiteSpace(text) ? null : text;
    }

    private string? ExtractFromText(byte[] content)
    {
        var text = Encoding.UTF8.GetString(content).Trim();
        return string.IsNullOrWhiteSpace(text) ? null : text;
    }
}
